src/main/java/online/book/store/service/book/BookServiceImpl.java

```diff
@@ -20,22 +25,31 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
+    private final CategoryRepository categoryRepository;
 
     @Override
     public BookResponseDto save(BookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
+        Set<Category> categories = requestDto.getCategoryIds()
+                .stream()
+                .map(categoryId
+                        -> categoryRepository.findById(categoryId)
+                        .orElseThrow(() -> new EntityNotFoundException(
+                                "Invalid category ID: " + categoryId)))
+                .collect(Collectors.toSet());
+        book.setCategories(categories);
```

Comment:
In this implementation you make N requests to db for each category, it is not efficient, better to take all existing categories dy id`s that comes with BookRequestDto  and then check size of obtained list